File: src/dependencies/crazyflie_hardware/src/crtp/crtp/packers/generic_commander_packer.py
from .packer import Packer
from .crtp_packer import CrtpPacker

import struct

from typing import Callable


class GenericCommanderPacker(Packer):
    PORT_COMMANDER_GENERIC = 0x07

    SET_SETPOINT_CHANNEL = 0
    META_COMMAND_CHANNEL = 1

    TYPE_STOP = 0
    TYPE_VELOCITY_WORLD = 1
    TYPE_ZDISTANCE = 2
    TYPE_HOVER = 5
    TYPE_FULL_STATE = 6
    TYPE_POSITION = 7

    TYPE_META_COMMAND_NOTIFY_SETPOINTS_STOP = 0

    def __init__(self, crtp_packer_factory: Callable[[int], CrtpPacker]):
        super().__init__(crtp_packer_factory, self.PORT_COMMANDER_GENERIC)

    def send_notify_sendpoints_stop(self, remain_valid_milliseconds):
        data = struct.pack(
            "<BI",
            self.TYPE_META_COMMAND_NOTIFY_SETPOINTS_STOP,
            remain_valid_milliseconds,
        )
        return self._prepare_packet(channel=self.META_COMMAND_CHANNEL, data=data)

    def send_stop_setpoint(self):
        data = struct.pack("<B", self.TYPE_STOP)
        return self._prepare_packet(channel=self.SET_SETPOINT_CHANNEL, data=data)

    def send_velocity_world_setpoint(self, vx, vy, vz, yawrate):
        data = struct.pack("<Bffff", self.TYPE_VELOCITY_WORLD, vx, vy, vz, yawrate)
        return self._prepare_packet(channel=self.SET_SETPOINT_CHANNEL, data=data)

    def send_zdistance_setpoint(self, roll, pitch, yawrate, zdistance):
        data = struct.pack(
            "<Bffff", self.TYPE_ZDISTANCE, roll, pitch, yawrate, zdistance
        )
        return self._prepare_packet(channel=self.SET_SETPOINT_CHANNEL, data=data)

    def send_hover_setpoint(self, vx, vy, yawrate, zdistance):
        data = struct.pack("<Bffff", self.TYPE_HOVER, vx, vy, yawrate, zdistance)
        return self._prepare_packet(channel=self.SET_SETPOINT_CHANNEL, data=data)

    def send_full_state_setpoint(
        self, x, y, z, vx, vy, vz, ax, ay, az, orient_comp, rr, pr, yr
    ):
        data = struct.pack(
            "<BhhhhhhhhhIhhh",
            self.TYPE_FULL_STATE,
            x,
            y,
            z,
            vx,
            vy,
            vz,
            ax,
            ay,
            az,
            orient_comp,
            rr,
            pr,
            yr,
        )
        return self._prepare_packet(channel=self.SET_SETPOINT_CHANNEL, data=data)

    def send_position_setpoint(self, x, y, z, yaw):
        data = struct.pack("<Bffff", self.TYPE_POSITION, x, y, z, yaw)
        return self._prepare_packet(channel=self.SET_SETPOINT_CHANNEL, data=data)

File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/packer/generic_commander_packer.cpp
#include "crtp_cpp/packer/generic_commander_packer.hpp"
#include <cstring>


#define PORT_COMMANDER_GENERIC 0x07

#define SET_SETPOINT_CHANNEL 0
#define META_COMMAND_CHANNEL 1

#define TYPE_STOP 0
#define TYPE_VELOCITY_WORLD 1
#define TYPE_ZDISTANCE 2
#define TYPE_HOVER 5
#define TYPE_FULL_STATE 6
#define TYPE_POSITION 7

#define TYPE_META_COMMAND_NOTIFY_SETPOINTS_STOP 0

GenericCommanderPacker::GenericCommanderPacker()
    : CrtpPacker(PORT_COMMANDER_GENERIC) {}

CrtpPacket GenericCommanderPacker::prepare_packet(const std::vector<uint8_t>& data, uint8_t channel) {
    return CrtpPacker::prepare_packet(channel, data);
}

CrtpPacket GenericCommanderPacker::send_notify_sendpoints_stop(uint32_t remain_valid_milliseconds) {
    std::vector<uint8_t> data(1 + 4);
    data[0] = TYPE_META_COMMAND_NOTIFY_SETPOINTS_STOP;
    std::memcpy(&data[1], &remain_valid_milliseconds, 4);
    return prepare_packet(data, META_COMMAND_CHANNEL);
}

CrtpPacket GenericCommanderPacker::send_stop_setpoint() {
    std::vector<uint8_t> data = {TYPE_STOP};
    return prepare_packet(data, SET_SETPOINT_CHANNEL);
}

CrtpPacket GenericCommanderPacker::send_velocity_world_setpoint(float vx, float vy, float vz, float yawrate) {
    std::vector<uint8_t> data(1 + 4 + 4 + 4 + 4);
    data[0] = TYPE_VELOCITY_WORLD;
    std::memcpy(&data[1], &vx, 4);
    std::memcpy(&data[5], &vy, 4);
    std::memcpy(&data[9], &vz, 4);
    std::memcpy(&data[13], &yawrate, 4);
    return prepare_packet(data, SET_SETPOINT_CHANNEL);
}

CrtpPacket GenericCommanderPacker::send_zdistance_setpoint(float roll, float pitch, float yawrate, float zdistance) {
    std::vector<uint8_t> data(1 + 4 + 4 + 4 + 4);
    data[0] = TYPE_ZDISTANCE;
    std::memcpy(&data[1], &roll, 4);
    std::memcpy(&data[5], &pitch, 4);
    std::memcpy(&data[9], &yawrate, 4);
    std::memcpy(&data[13], &zdistance, 4);
    return prepare_packet(data, SET_SETPOINT_CHANNEL);
}

CrtpPacket GenericCommanderPacker::send_hover_setpoint(float vx, float vy, float yawrate, float zdistance) {
    std::vector<uint8_t> data(1 + 4 + 4 + 4 + 4);
    data[0] = TYPE_HOVER;
    std::memcpy(&data[1], &vx, 4);
    std::memcpy(&data[5], &vy, 4);
    std::memcpy(&data[9], &yawrate, 4);
    std::memcpy(&data[13], &zdistance, 4);
    return prepare_packet(data, SET_SETPOINT_CHANNEL);
}

CrtpPacket GenericCommanderPacker::send_full_state_setpoint(int16_t x, int16_t y, int16_t z, int16_t vx, int16_t vy, int16_t vz, int16_t ax, int16_t ay, int16_t az, uint32_t orient_comp, int16_t rr, int16_t pr, int16_t yr) {
    std::vector<uint8_t> data(1 + 2 + 2 + 2 + 2 + 2 + 2 + 2 + 2 + 2 + 4 + 2 + 2 + 2);
    data[0] = TYPE_FULL_STATE;
    std::memcpy(&data[1], &x, 2);
    std::memcpy(&data[3], &y, 2);
    std::memcpy(&data[5], &z, 2);
    std::memcpy(&data[7], &vx, 2);
    std::memcpy(&data[9], &vy, 2);
    std::memcpy(&data[11], &vz, 2);
    std::memcpy(&data[13], &ax, 2);
    std::memcpy(&data[15], &ay, 2);
    std::memcpy(&data[17], &az, 2);
    std::memcpy(&data[19], &orient_comp, 4);
    std::memcpy(&data[23], &rr, 2);
    std::memcpy(&data[25], &pr, 2);
    std::memcpy(&data[27], &yr, 2);
    return prepare_packet(data, SET_SETPOINT_CHANNEL);
}

CrtpPacket GenericCommanderPacker::send_position_setpoint(float x, float y, float z, float yaw) {
    std::vector<uint8_t> data(1 + 4 + 4 + 4 + 4);
    data[0] = TYPE_POSITION;
    std::memcpy(&data[1], &x, 4);
    std::memcpy(&data[5], &y, 4);
    std::memcpy(&data[9], &z, 4);
    std::memcpy(&data[13], &yaw, 4);
    return prepare_packet(data, SET_SETPOINT_CHANNEL);
}